### ISSUE TITLE
Add polling API with setting

### ISSUE THREAD

**Issue by mauteri** [GatherPress/gatherpress#428]

**Comment by mauteri** [comment]
> There is no need for a polling api, the standard way of doing this in wordpress is the heartbeat api just piigy back on that

Isn't heartbeat API only in the admin? I might be wrong, but thought it was just an admin thing and couldn't be used on the frontend, where it would be needed here.

**Comment by shawfactor** [comment]
Core uses it in only back end and with the admin bar. But it is used by 100s of plugins in the front end and the Wordpress documentation specifically mentions the front end

https://developer.wordpress.org/plugins/javascript/heartbeat-api/

I use it extensively in my plugins and of course it’s used by buddypress/buddyboss.


It is the standard way of doing polling and the beauty of it is that you add your requests to the heartbeat which may already be running so there are no additional http requests and meaning lower server load than rolling your own solution

**Comment by mauteri** [comment]
Ok, I haven't really used it that much and thought it was very much an admin thing. I was under the impression it leveraged admin-ajax which isn't something you would want to use on the frontend. I'll take a look. Thanks for the info!

**Comment by shawfactor** [comment]
It does use admin-ajax.php. But Im' not sure what the issue is using that on the front end?

One day wordpress will add similar functionality to the json api but for now heartbeat is the standard way of doing two way polling in wordpress both on the front and back end.

The major advantages are it has a good api (no need to reinvent the wheel) and because you piggy back on the existing polling heartbeat you are not creating extraneous http requests

**Comment by carstingaxion** [comment]
@mauteri @shawfactor Is this still needed or wanted, now that GatherPress uses the Interactivity API all over the place?

**Comment by mauteri** [comment]
@carstingaxion Still think it's a pretty cool feature, so I'd keep it open.

It's a different thing from the Interactivity API, not something it replaces. The Interactivity API updates the UI based on what *you* do — you RSVP, your button updates. What it doesn't do is pick up changes other people make on the server: someone else RSVPs and the count goes up, the waiting list shifts, an online event link becomes available. You won't see any of that until you reload the page.

That's what this issue is about — checking the server every so often so the page reflects what's changed elsewhere, for a more real-time feel.

If anything, now that we use the Interactivity API everywhere, this gets easier than the original plan. Instead of the old admin-ajax/Heartbeat approach, a simple REST poll could feed the interactivity store and the UI updates on its own. Still want it as a setting that's off by default, since polling on an interval can be rough on some shared hosts.

So yeah, keep it open, but worth revisiting how we'd build it now.

**Comment by carstingaxion** [comment]
Aha, now that I understand,  it makes a lot of sense to keeping.

Thanks for the comprehensive explanation @mauteri !

**Comment by shawfactor** [comment]
Yes but the standard for all Wordpress polling is the heartbeat api and if
it’s already running you can just piggy back on it (ie no extra http
requests)

**Comment by mauteri** [comment]
Direction for a v1: start with the **Heartbeat API, enqueued only for logged-in users**.

**Why logged-in only:** Heartbeat ticks are POSTs to `admin-ajax.php` — uncached by design, and they boot WordPress fully each time. For logged-in users that's fine; they're already getting uncached pages and they're a small audience (and the ones who benefit). For anonymous visitors it's the opposite — those pageviews normally come straight from page cache/CDN with no PHP, and firing Heartbeat for them turns cheap cached views into a stream of uncached requests. On a busy event page that's a load spike, which is the shared-hosting concern from the top of this issue. So gating on `is_user_logged_in()` is the fix.

It also lines up with what the feature actually needs — the online event link (logged-in + attending) and RSVP updates are logged-in actions anyway. The only thing an anonymous viewer misses is the live attendee count ticking up, and they still get that on page load. Realistically this stays a logged-in feature: the payload is per-user (your RSVP status, whether you can see the online link), so there's nothing to cache and Heartbeat's uncached nature is a non-issue.

**Rough plan:**
- Enqueue `heartbeat` on the frontend only when logged in, and only on views that need it (single event / pages with an event or RSVP block) — not site-wide.
- Server side, hook the `heartbeat_received` filter to attach the GatherPress data (RSVP counts, current user's status, online link) when the tick carries an event ID. Keep it cheap since it runs per tick per viewer.
- Bump the interval to ~30–60s via the `heartbeat_settings` filter (default ~15s is more than we need; it already backs off when idle and pauses on hidden tabs).
- The tick handler just dispatches into the Interactivity store and the UI updates on its own.
- Keep it behind the off-by-default setting.

**Comment by carstingaxion** [comment]
I like the idea to decrease the heartbeat interval, as I've done this before too.
But I don't think the setting should be changed by GatherPress itself, even it makes sense for our use-case.

Imagining that I've a website, that heavily relies on the heartbeat (with default or custom set interval): It could become a not-nice developer-experience to debug this side effect. 

Why not: Just put a note next to the toggle in settings, that GatherPress would still run smoothly, when the heartbeat interval would be decreased.
